File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.cpp
class Solution {
  public:
    int countSubstr(string& s, int k) {
        // code here.
        int n=s.size();
        vector<int>mp(26,0);
        int j=0;
        int ans=0;
        int count=0;
        vector<int>dp(26,-1);
        for(int i=0;i<n;i++)
        {
            mp[s[i]-'a']++;
            dp[s[i]-'a']=i;
            if(mp[s[i]-'a']==1)
            {
                ++count;
            }
            while(j<i && count>k )
            {
                --mp[s[j]-'a'];
                if(mp[s[j]-'a']==0)
                {
                    --count;
                }
                j++;
            }
            if(count<k)
            {
                continue;
            }
            int mini=INT_MAX;
            for(int l=0;l<26;l++)
            {
                
                if(dp[l]>=j)
                {
                  mini=min(mini,dp[l]);  
                }
            }
            ans+=(mini-j+1);
        }
        return ans;
    }
};

File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.java
class Solution {
   public int countSubstr(String s, int k) {
    return atMostK(s, k) - atMostK(s, k - 1);
}

private int atMostK(String s, int k) {
    if (k < 0) {
        return 0;
    }

    int[] freq = new int[26];
    int left = 0;
    int distinct = 0;
    int ans = 0;

    for (int right = 0; right < s.length(); right++) {
        int c = s.charAt(right) - 'a';
        if (freq[c] == 0) {
            distinct++;
        }
        freq[c]++;

        while (distinct > k) {
            int lc = s.charAt(left) - 'a';
            freq[lc]--;
            if (freq[lc] == 0) {
                distinct--;
            }
            left++;
        }

        ans += right - left + 1;
    }

    return ans;
}
}